apm-agent-core/src/main/java/co/elastic/apm/agent/impl/circuitbreaker/CircuitBreaker.java

```diff
@@ -0,0 +1,144 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.impl.circuitbreaker;
+
+import co.elastic.apm.agent.context.AbstractLifecycleListener;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.util.ExecutorUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class CircuitBreaker extends AbstractLifecycleListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(CircuitBreaker.class);
+
+    private final ThreadPoolExecutor threadPool;
+    private final ElasticApmTracer tracer;
+    private final CircuitBreakerConfiguration circuitBreakerConfiguration;
+    private final long pollInterval;
+
+    private boolean isCurrentlyUnderStress = false;
+
+    private final List<StressMonitor> stressMonitors = new CopyOnWriteArrayList<>();
+
+    public CircuitBreaker(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        circuitBreakerConfiguration = tracer.getConfig(CircuitBreakerConfiguration.class);
+        pollInterval = circuitBreakerConfiguration.getStressMonitoringPollingInterval();
+        threadPool = ExecutorUtils.createSingleThreadDeamonPool("circuit-breaker", 1);
+    }
+
+    @Override
+    public void start(ElasticApmTracer tracer) {
+        // failsafe loading of stress monitors in isolation
+        loadGCStressMonitor(tracer);
+
+        threadPool.execute(new Runnable() {
+            @Override
+            public void run() {
+                pollStressMonitors();
+            }
+        });
+    }
+
+    private void loadGCStressMonitor(ElasticApmTracer tracer) {
+        try {
+            stressMonitors.add(new GCStressMonitor(tracer));
+        } catch (Throwable throwable) {
+            logger.error("Failed to load the GC stress monitor. Circuit breaker will not be triggered based on GC events.", throwable);
+        }
+    }
+
+    private void pollStressMonitors() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                if (circuitBreakerConfiguration.isCircuitBreakerEnabled()) {
+                    if (isCurrentlyUnderStress) {
+                        boolean stressRelieved = true;
+                        for (StressMonitor stressMonitor : stressMonitors) {
```

Comment:
this is a bit hard to read due to lots of conditionals & try/catch, maybe splitting in two smaller methods `isStressRelieved` and `isUnderStress` would make the intent a bit easier to grasp.